### PR TITLE
fix the math module for autoray 0.7.0

### PR DESCRIPTION
**Pull request by albi3ro**
**Context:**

Autoray has released a new version. This has caused failures with several tensorflow functions.

**Description of the Change:**

Patches tensorflow functions that are now longer working.

`null.qubit` no longer returns interface numbers unless we have `gradient_method="backprop"`. This is necessary because null qubit was creating tensorflow arrays with the wrong precision.  A device does not need to return interface data in any situation other than backprop.  That conversion to the interface data type was unnecessary.

**Benefits:**

No more autoray version pinning.

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/math/single_dispatch.py [hunk 1]
```diff
@@ -267,11 +267,12 @@ def _take_autograd(tensor, indices, axis=None):
...
-    return ar.autoray.tensorflow_diag(_tf_convert_to_tensor(x), **kwargs)
-
-
-ar.register_function("tensorflow", "diag", _coerce_tensorflow_diag)
+ar.register_function(
+    "tensorflow", "unstack", lambda *args, **kwargs: _i("tf").unstack(*args, **kwargs)
+)
+ar.register_function(
+    "tensorflow", "gather", lambda *args, **kwargs: _i("tf").gather(*args, **kwargs)
+)
```
Comment (mudit2812) [inline pennylane/math/single_dispatch.py:275]: I don't understand why this does not work without manually registering the functions 😕

File: tests/math/test_functions.py [hunk 2]
```diff
@@ -2953,3 +2953,11 @@ def jitted_function(y):
...
         assert qml.math.allclose(array2, jnp.array([[7, 2, 3, 4]]))
         assert isinstance(array2, jnp.ndarray)
+
+
+def test_unstack_tensorflow():
+    """Test that unstack works with tensorflow variables."""
+    x = tf.Variable([0.1, 0.2])
+    r1, r2 = qml.math.unstack(x)
+    assert qml.math.allclose(r1, tf.Variable(0.1))
+    assert qml.math.allclose(r2, tf.Variable(0.2))
```
Comment (mudit2812) [inline tests/math/test_functions.py:2963]: Could you also add a test for `gather`?

**Review by mudit2812 — COMMENTED** [review]
Looks good.

> We will no longer work with autoray < 0.7.0.

Why will autoray<0.7 stop working? We've only registered 2 new functions.

File: tests/math/test_functions.py (hunk 2 above)
Comment (albi3ro) [reply tests/math/test_functions.py:2963]: There are pre-existing tests in this file up above.

File: pennylane/math/single_dispatch.py (hunk 1 above)
Comment (albi3ro) [reply pennylane/math/single_dispatch.py:275]: I think autoray is now looking in the experiental module for them.

Comment (mudit2812) [reply pennylane/math/single_dispatch.py:288]: Silly `autoray`

**Comment by albi3ro** [comment]
[sc-76593]

**Review by mudit2812 — APPROVED** [review]

File: pennylane/math/single_dispatch.py (hunk 1 above)
Comment (dwierichs) [reply pennylane/math/single_dispatch.py:288]: I think we may want to do those with 
```python
ar.autoray._SUBMODULE_ALIASES["tensorflow", "unstack"] = "tensorflow"
ar.autoray._SUBMODULE_ALIASES["tensorflow", "gather"] = "tensorflow"
ar.autoray._SUBMODULE_ALIASES["tensorflow", "concat"] = "tensorflow"
```
like about 40 lines above. :) Also seems like we can remove some the `tensorflow.experimental.numpy` registrations up there, now that autoray looks in that module anyways.

**Review by dwierichs — APPROVED** [review]
Only had an optional suggestion to simplify the function registration for non-experimental tensorflow functions, and the idea to remove now redundant pointers to the experimental module that we added earlier.
Thanks for fixing this!

**Review by PietropaoloFrisoni — APPROVED** [review]

File: pennylane/devices/null_qubit.py [hunk 3]
```diff
@@ -401,16 +416,24 @@ def compute_vjp(
...
         cotangents: tuple[Number],
         execution_config: ExecutionConfig = DefaultExecutionConfig,
     ):
-        results = tuple(
-            self._simulate(c, INTERFACE_TO_LIKE[execution_config.interface]) for c in circuits
+        interface = (
+            INTERFACE_TO_LIKE[execution_config.interface]
+            if execution_config.gradient_method == "backprop"
+            else "numpy"
         )
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/devices/null_qubit.py:436]: extracting a helper function.

File: pennylane/math/single_dispatch.py (hunk 1 above)
Comment (albi3ro) [reply pennylane/math/single_dispatch.py:288]: submodule aliases are workinng, so I'll use that instead.

**Pull request merged**